[Navigation screen] Separate block navigator focus from the editor focus

adamziel opened the pull request:
## Description

Closes #22705

This PR is an attempt to separates the block navigator selection/focus from the editor area. Once applied, all actions performed in the navigator will only affect selection/focus within the navigator instead of trying to synchronize with the editor area:

![2020-06-08 16-05-56 2020-06-08 16_06_55](https://user-images.githubusercontent.com/205419/84039888-322fa200-a9a2-11ea-9a8d-27f4f9048f8d.gif)

CC @enriquesanchez and @karmatosed for sanity check

Editing the caption is not implemented yet.

Open questions:
* Does it feel like the right approach?
* The concept of selecting a navigator item, as in turning the icon dark once an item is clicked, does not seem to be useful anymore - should we get rid of it and switch to the caption edit mode on click?

## How has this been tested?
1. Enable the navigation experiment in Gutenberg > Experiments
1. Go to Gutenberg > Navigation (beta) and create a menu
1. Use the navigator using a keyboard, select/remove/duplicate some navigation items, confirm the behavior makes intuitive sense
1. Go to the post editor, create a navigation block, select/move some navigation items using the navigator, confirm the behavior did not change.

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamziel committed:
Separate the navigator selection from the editor area selection

adamziel committed:
add __experimental prefix and use list of selected blocks instead of a single id

adamziel committed:
Don't attempt to call __experimentalSelectBlock when it's not provided

adamziel committed:
Drop onClick and only pass selectBlock to BlockNavigationBlock

adamziel commented:
This PR breaks some tests ATM, I'd like to get some initial feedback on direction before debugging them.

talldan commented:
I agree with @enriquesanchez that is feels a lot more reliable and obvious where focus will end up. 👍 

I'm not sure what to do about the selected block being highlighted, potentially it's not needed any more, although at the same time it might be useful for a user looking to jump between the normal editor and the block navigation. Though it could also be confusing.

I notice we're showing the movers and settings menu for the selected block, which is probably not needed now that you can't select the block in the block navigation. I think they should be shown for the focused and hovered blocks only.

I still think an option to jump to the actual block from the block navigation might be useful, I wonder if we might want to explore a 'Go To ...' option in the settings menu as part of a separate issue?

draganescu commented:
I just want to +1 the "Go to..." idea!
Options for the copy:
- Select in menu
- Find in menu
- Go to item
- Show item

enriquesanchez commented:
>I'm not sure what to do about the selected block being highlighted, potentially it's not needed any more, although at the same time it might be useful for a user looking to jump between the normal editor and the block navigation. Though it could also be confusing.

I was actually confused and did not know that the block icon was meant to take you to the nav block. I don't think that's obvious, we certainly don't have that same interaction anywhere else, do we? 

>I still think an option to jump to the actual block from the block navigation might be useful, I wonder if we might want to explore a 'Go To ...' option in the settings menu as part of a separate issue?

+1 to this idea, but I'm finding it hard to come up with the right text for this. In addition to Andrei's suggestions, I'll add "Go to block"

adamziel committed:
Remove selection CSS

adamziel committed:
Stop rendering movers for selected block

adamziel committed:
Add "Go to block" navigator menu action

adamziel commented:
All good notes! I removed the dark CSS when an item is "selected" and added a "Go to" action (naming TBD):

<img width="488" alt="Zrzut ekranu 2020-06-12 o 17 55 32" src="https://user-images.githubusercontent.com/205419/84521918-04b26380-acd6-11ea-9e92-d3eb22ede550.png">

The next step here for me is to fix unit tests and rigorously user-test the entire PR.

draganescu commented:
I love that Go to option! Feels pretty neat :)
I tested this locally and in Firefox it seems that some of the clicks did not register correctly:

![weird focus](https://user-images.githubusercontent.com/107534/84649047-88569500-af0e-11ea-868f-f1e61893ffe2.gif)

adamziel committed:
Fix unit test

adamziel committed:
Restore the previous way of focusing items in the navigator

adamziel committed:
Lint

adamziel committed:
Ensure clicking go to block always works

adamziel commented:
@draganescu good catch! I just fixed the non-responding "Go to block" button. I am unable to reproduce the other issue (non-responding navigator item). Let's get these tests green and see if it's still happening.

adamziel commented:
Alright, this one seems to be ready for re-review!

draganescu commented on packages/block-editor/src/components/block-navigation/block.js:
```diff
@@ -144,7 +157,24 @@ export default function BlockNavigationBlock( {
+										onClick={ async () => {
+											// If clientId is already selected, it won't be focused (see block-wrapper.js)
+											// This removes the selection first to ensure the focus will always switch.
+											await selectEditorBlock( null );
```
This seems .... weird. I understand from the comment what it does but the presence of this line feels like something is not happening right.

adamziel replied:
@draganescu I agree, at the same time I don't think it's a blocking issue - I'd love to follow up with a better solution once this PR lands. What do you think?

draganescu replied:
But do we know what that followup will be about?

adamziel replied:
@draganescu the root cause is this effect here:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/block-wrapper.js#L136-L140

As you can see, it applies focus when `isSelected` changes. If the block loses focus but not selection, selecting it again will not trigger that effect. I don't know what a good solution would look like just yet though - I need to give it more thought.

draganescu replied:
Added #23269 to track this!

adamziel committed:
Add a README file

draganescu reviewed (approved):
I tested this locally and it works to keep the selection in the navigator for the navigation screen while not affecting how selection works in the post editor.

There is a spinoff issue #23269 that should improve how returning the focus to the block list works, as right now a nested block doesn't get it's editable area focused.
